cmd/relay-metrics/main.go

```diff
@@ -0,0 +1,275 @@
+	err := c.List(ctx, wrs)
+	if err != nil {
+		panic(err.Error())
+	}
+	for _, item := range wrs.Items {
+		m := workflowRunMetric{
+			Name:              item.Name,
+			Status:            item.Status.Status,
+			SecondsSinceStart: now.Sub(item.ObjectMeta.CreationTimestamp.UTC()).Seconds(),
+		}
+		fmt.Printf("%s is %s, %d\n", m.Name, m.Status, int(m.SecondsSinceStart))
+		ret = append(ret, &m)
+	}
+	return
+}
+
+func homeDir() string {
+	if h := os.Getenv("HOME"); h != "" {
+		return h
+	}
+	return os.Getenv("USERPROFILE") // windows
+}
+
+func k8sClient(kubeconfig *string) client.Client {
+	// Try homedir config
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
+	if err != nil {
+		// Try in-cluster config
+		config, err = rest.InClusterConfig()
+		if err != nil {
+			panic(err.Error())
+		}
+	}
+
+	c, err := client.New(config, client.Options{
+		Scheme: Scheme,
+	})
+	if err != nil {
+		panic(err.Error())
+	}
+
+	return c
+}
+
+func metricsClient(ctx context.Context) *monitoring.MetricClient {
+	mc, err := monitoring.NewMetricClient(ctx)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	return mc
+}
+
+func main() {
+	ctx := context.Background()
+	var err error
+	var kubeconfig *string
+
+	if home := homeDir(); home != "" {
+		kubeconfig = flag.String("kubeconfig", filepath.Join(home, ".kube", "config"), "(optional) absolute path to the kubeconfig file")
+	} else {
+		kubeconfig = flag.String("kubeconfig", "", "absolute path to the kubeconfig file")
+	}
+	publish := flag.String("publish", "false", "Publish metrics to stackdriver")
+	environment := flag.String("environment", "testing", "The environment being monitored for workflow runs")
+	deleteMetrics := flag.String("delete", "false", "Delete the metric descriptors. Be very sure")
+
+	flag.Parse()
+
+	kc := k8sClient(kubeconfig)
+	mc := metricsClient(ctx)
+
+	statusName := "Pending Workflow Runs"
+	oldestName := "Oldest Pending Workflow Run"
+	statusType := "custom.googleapis.com/relay/workflowruns/status"
+	oldestType := "custom.googleapis.com/relay/workflowruns/oldest_pending"
+
+	if *deleteMetrics == "true" {
+		err = deleteMetric(os.Stdout, "projects/"+projectID+"/metricDescriptors/"+statusType)
+		if err != nil {
+			panic(err.Error())
+		}
+		err = deleteMetric(os.Stdout, "projects/"+projectID+"/metricDescriptors/"+oldestType)
+		if err != nil {
+			panic(err.Error())
+		}
+		return
+	}
+
+	if publish != nil && *publish == "true" {
+		_, err = createCustomMetric(ctx, mc, os.Stdout, statusName, statusType, "The number of Workflow Runs with a status of `pending`")
+		if err != nil {
+			panic(err.Error())
+		}
+
+		_, err = createCustomMetric(ctx, mc, os.Stdout, oldestName, oldestType, "The number of seconds since the oldest Workflow Runs with a status of `pending` was started")
+		if err != nil {
+			panic(err.Error())
+		}
+	}
+
+	for {
+		statuses := getStatuses(ctx, kc)
+		count := 0
+		oldest := 0.0
+		for _, status := range statuses {
+			if status.Status == "pending" {
```

Comment:
You also want to check for no status at all to be careful.

Reply:
We have four workflow runs stuck in staging with no status. The longest has been around for 2356830 seconds (about 27 days).

Reply:
That's okay! We need to fix them or delete them. We should be alerted when they get stuck like that.